Dpg and ApiView integration doc (#19122)
Azure/azure-rest-api-specs: documentation/onboard-dpg-in-sdkautomation/Integrate-dpg-and-apiview-in-sdk-automation-pipeline.md

# Integrate DPG and ApiView in SDK Automation Pipeline
This document is targeting for everyone who wants to know all the implementation details of SDK automation pipeline.
It describes the details of integrating DPG and ApiView in SDK Automation Pipeline.

Before go through this document, please go through [Service Onboard DPG with Swagger CI Pipeline](README.md) to be familiar with the user experience in service team side. Also, you need to be familiar with [SDK Automation Pipeline Framework](../sdkautomation/README.md).

# WorkFlow
![integrate-dpg-and-apiview](integrate-dpg-and-apiview.png)

__Description:__
1. SDK Automation Pipeline is triggered in spec PR CI.

> 2. SDK Automation Pipeline Framework checks whether there is a comment containing autorest configuration in spec PR. If found, generate sdk by the autorest configuration in comment.

Author: automation script will generate autorest.md and use the generated autorest.md to generate sdk.

About merge the autorest configuration in pr comment and existing autorest.md, I think replacing the existing one is prefered by sdk owners because we may encounter unexpected error in merging autorest configuration for multi client use case.

# Integrate DPG and ApiView in SDK Automation Pipeline
This document is targeting for everyone who wants to know all the implementation details of SDK automation pipeline.
It describes the details of integrating DPG and ApiView in SDK Automation Pipeline.

Before go through this document, please go through [Service Onboard DPG with Swagger CI Pipeline](README.md) to be familiar with the user experience in service team side. Also, you need to be familiar with [SDK Automation Pipeline Framework](../sdkautomation/README.md).

# WorkFlow
![integrate-dpg-and-apiview](integrate-dpg-and-apiview.png)

__Description:__
1. There is swagger PR triggering the SDK Automation Pipeline.
2. SDK Automation Pipeline Framework checks whether there is a branch named `dpg/<prNumber>` in sdk repository. If yes, the branch `dpg/<prNumber>` will be used as base branch to generate SDK. Otherwise, `main` branch will be used as base branch.
   1. `<prNumber>` is the swagger PR number.

>    2. If there is branch `dpg/<prNumber>`, `main` branch is usually used and sometimes not. It depends on the [specificationRepositoryConfiguration](../../specificationRepositoryConfiguration.json).

Reviewer: could you describe more about this?

Author: I add decription of which configuration is used.

> 3. SDK automation script finds the corresponding `autorest.md` or `README.md` in sdk repository. If there is no `autorest.md` or `README.md` founded, please output `{"packages": []}` in `generateOutput.json`, then the pipeline finishes and skipp the following steps.

Reviewer: ```suggestion
3. SDK automation script will search for the corresponding `autorest.md` or `README.md` in sdk repository. If those files are not found, the pipeline will  output `{"packages": []}` in `generateOutput.json`, stop the generation.
```

>    1. A relative readme.md and changedFiles of the swagger PR can be get in `generateInput.json`, and automation script can use it to find the corresponding `autorest.md` and `README.md`.

Reviewer: Better to use configration files instead of `autorest.md` and `README.md`.

   2. We will only use `require` to include swagger readme.md in `autorest.md` or `README.md`, and `input-file` should not be used in `autorest.md` or `README.md` in most times, but there are some exceptions:
      1. .Net SDK may continue to use `input-file` because existing HLC dataplane sdk may already use `input-file`.
      2. Multi-Client for Java, Python and JS cannot use `readme.md` in swagger directly because they use `batch` task, so using `input-file` is better.

> 4. SDK automation script modifies the founded `autorest.md` or `README.md`.

Reviewer: ```suggestion
4. SDK automation script will modify the configuration files in the SDK repo`.
```
Could you explain why?

   1. Change the `require` block to include the latest swagger `readme.md` in the PR. The value can be joined by `specFolder` and `relatedReadmeMdFiles` in `generateInput.json`. For example:
      ```yaml
      require:
        - ../../../../../azure-rest-api-specs/specification/deviceupdate/data-plane/readme.md
      ```
   2. For exceptions which still use `input-file`, replace the value of `input-file` to include the latest swagger in the PR. The value can be calculated same as `require` block.

> 5. SDK automation script generates SDK with the modified `autorest.md` or `README.md`.

Reviewer: Put 5,6,7 together?  

DK automation script will generate SDK from the updated configuration files, create APIViews and integrate those information into the Swagger PR.

# Integrate DPG and ApiView in SDK Automation Pipeline
This document is targeting for everyone who wants to know all the implementation details of SDK automation pipeline.

> It describes the details of integrating DPG and ApiView in SDK Automation Pipeline.

Reviewer 2: Note that the existence of ApiView for review depends on service onboarding DPG on day one.

I am a bit worried that service does not aware of DPG and hence do not even decide to onboard, when they create the swagger and ask for API review.
Later when they request the SDK for their GA, rejecting them is not an option, and hence DPG is still required, even maybe months after the API review (another alternative would be HLC by service themselves).

Reviewer 3: I suppose this depends on when you start counting days. I think that service teams often create PRs in the azure-rest-api-specs repo and then work on them for days, weeks, and sometimes months before the API review is scheduled. At any point in this process the service team can create whatever configuration is needed to produce DPG client libraries so that they can be used in the REST API review.

But more importantly, it's not clear to me that DPG libraries _need_ to be available in the very initial REST API reviews of a new service. Often the initial review for a service's REST API focuses on fundamental issues of resource design and hierarchy, path structure, LRO patterns, and other details purely at the REST API level.  I don't think DPG libraries will be useful for these early reviews. They will be useful later, when issues of naming and client structure are considered -- but that is the point when client teams _have_ to be actively engaged in   producing client libraries.  Note that APEX _requires_ SDKs for a public preview -- well before any GA -- so there is little chance that a service team will get to the point of wanting to GA without already having produced and released at least one version of their client libraries.

Author: Hi @mikekistler What we want to do is generating DPG libraries after the REST API review and before the REST API PR is merged. It's not reviewing DPG libraries and REST API at the same time. The advantage is that service team can update the swagger PR directly when there is some comments from Arch Board Review, insteading of creating a new PR to update swagger.

Reviewer 3: Maybe we have different goals then, because "reviewing DPG libraries and REST API at the same time" is _exactly_ what I want to do.  I believe this combined review is the mechanism to identify the changes needed in the REST API PR.

Author: Get it. When the DPG is integrated into sdk automation pipeline, I think we also can review DPG libraries and REST API at the same time

Reviewer 2: @mikekistler 
Yes, I agree that the "production ready" quality of the SDK would be likely be required at service public preview, and certainly at their GA.

What I mean is that
1. When service do their first API review, they may not have good idea on who is going to build their SDK (SDK team? Self-serve via DPG? Service team?) and how it is built. (Ted's WIP on join session with SDK arch board and ARM should largely solve it).
2. Even if service decide to build SDK with their own resource (i.e., don't do DPG at all), a generated SDK would still be helpful, as reviewer can see the API in a different (maybe familiar) perspective (I actually prefer traditional SDK with models, instead of DPG on this purpose). I think yours and Wei's latter comment is of the similar intention.

A far-fetched example: I was aware very recently that Purview Scanning had a big problem with their swagger on all api-versions (after sign-off and PR merged).
https://github.com/Azure/azure-rest-api-specs/pull/17382#issuecomment-1134327699
I found this only when I am testing the SDK codegen to generate serialization/de-serialization unit test with mock data (so it requires model and CI run).

Before go through this document, please go through [Service Onboard DPG with Swagger CI Pipeline](README.md) to be familiar with the user experience in service team side. Also, you need to be familiar with [SDK Automation Pipeline Framework](../sdkautomation/README.md).

# WorkFlow

> ![integrate-dpg-and-apiview](integrate-dpg-and-apiview.png)

Reviewer 3: I recommend you checkout [mermaid.js](https://mermaid-js.github.io/mermaid/#/). GitHub supports mermaid.js natively so there is no need to check in .png files, which means a year from now no one has to hunt down the source and tools used to create the .png.

Author: I maybe not a good mermaid user. When I try to migrate the workflow to mermaid, the flowchart rendered by mermaid is not easy for read. The following is the flowchart:
```mermaid
flowchart TD
   serviceTeam(Service Team) --> swaggerPR(Swagger PR)
   swaggerPR --> choose1{Have you added a comment containing autorest configuration\n in spec PR and run command `/azp run`?}
   choose1 -->|Yes| generate1(Use the information in comment to generate sdk and ApiView)
   generate1 --> choose2{Is API Definition Approvied?}
   choose2 -->|Yes| END(End)
   choose2 -->|No| change(Add/Update the comment that contains autorest configuraiton, and then re-run pipeline by command `/azp run`)
   change --> generate1
   choose1 -->|No| choose3{Is there autorest configuration file in sdk repo?}
   choose3 -->|No| END
   choose3 -->|Yes| generate2(Generate sdk by using autorest configuration in sdk repo)
   generate2 --> choose2
```

__Description:__
1. There is swagger PR triggering the SDK Automation Pipeline.

> 2. SDK Automation Pipeline Framework checks whether there is a branch named `swagger/<prNumber>` in sdk repository. If yes, the branch `swagger/<prNumber>` will be used as base branch to generate SDK. Otherwise, `main` branch will be used as base branch.

Reviewer 3: Can we please name the branch `restapi/<prNumber>`? I know folks in Microsoft love the name "swagger" but it was changed to "OpenAPI V2" over five years ago.

Author: sure. But Wes proposed a new idea in the email thread, and I think we don't need this branch if we use the new idea.

> 2. SDK Automation Pipeline Framework checks whether there is a branch named `swagger/<prNumber>` in sdk repository. If yes, the branch `swagger/<prNumber>` will be used as base branch to generate SDK. Otherwise, `main` branch will be used as base branch.

Author: yes

   1. `<prNumber>` is the swagger PR number.
   2. If there is branch `swagger/<prNumber>`, `main` branch is usually used and sometimes not. It depends on the `mainRepository` defined in [specificationRepositoryConfiguration](../../specificationRepositoryConfiguration.json).

> 3. SDK automation script will search for the corresponding autorest configuration file in sdk repository. If those files are not found, the pipeline automation script will output `{"packages": []}` in `generateOutput.json`, and stop the generation.

Reviewer 4: In the case were we don't find an autorest.md we should consider adding a message with link to instructions on how to get a SDK generated.

Author: please review the latest doc.

   1. A relative readme.md and changedFiles of the swagger PR can be get in `generateInput.json`, and automation script can use it to find the corresponding autorest configuration file in sdk repository.
   2. We will only use `require` to include swagger readme.md in autorest configuration file, and `input-file` should not be used in autorest configuration file in most times, but there are some exceptions:
      1. .Net SDK may continue to use `input-file` because existing HLC dataplane sdk may already use `input-file`.
      2. Multi-Client for Java, Python and JS cannot use `readme.md` in swagger directly because they use `batch` task, so using `input-file` is better.
4. SDK automation script will modify the autorest configuration file in the SDK repo`.
   1. Change the `require` block to include the latest swagger `readme.md` in the PR. The value can be joined by relative path from package folder to sdk root, `specFolder` and `relatedReadmeMdFiles` in `generateInput.json`. For example:
      ```yaml
      require:

>         - ../../../../../azure-rest-api-specs/specification/deviceupdate/data-plane/readme.md

Reviewer 4: Should this instead be a link to the sha of from the PR instead of a relative path? If we end up needing to submit this change to the SDK language repos we will want it to be sha based.

Author: No. In step of generating sdk in sdk automation pipeline, we cannot get the link to the sha because the PR is not merged, and we cannot get the link to the final version of merged one.

      ```
   2. For exceptions which still use `input-file`, replace the value of `input-file` to include the latest swagger in the PR. The value can be calculated similar as `require` block.
5. SDK automation pipeline generates SDK and ApiView with the modified autorest configuration file , and then add comments about results to the Swagger PR.

# Future Work

> Currently, we ask service team to create branch `swagger/<prNumber>` and add/update autorest configuration file manually because the configuration file includes some necessary information to generate SDK.

Reviewer 4: If we go with the new idea I was outlining then in the future in the PowerApps workflow for creating the specs PR we would also have the service teams provide the other necessary metadata and we would include that in whatever format we come up with in the spec PR creation.

Author: yes.
